### PR TITLE
IStorageProvider API updates

### LINKED ISSUE #9029: v11 Open/Save File API Suggestions
I'm looking to mirror Avalonia's new API for Open/Save File/Folders into MvvmDialogs, but there are a few design decisions I'd like to question.

IStorageItem is meant to be a consumption-only API, but there is one location where it needs to be set: `PickerOptions.SuggestedStartLocation`. How is that property supposed to be set?? I don't even know.

Why not simply have a direct nullable Uri property on `IStorageItem`? If you don't want to resolve it automatically for performance reason, it can be lazy-loaded, and the `bool` value is simply repeating whether the value is null or not. I see no benefit to this function and it makes it more complex to use.

```c#
/// <summary>
/// Gets the full file-system path of the item, if the item has a path.
/// </summary>
/// <remarks>
/// Android backend might return file path with "content:" scheme.
/// Browser and iOS backends might return relative uris.
/// </remarks>
bool TryGetUri([NotNullWhen(true)] out Uri? uri);
```

I'll replace this in my API with
```c#
Uri? Path { get; }
```

### LINKED ISSUE #10034: Further improve new FilePicker API
The new FilePicker API is great but also a bit confusing and annoying. For example:

##### Problem:
`FilePickerOpenOptions` and `FilePickerSaveOptions` both have file type filters. But the naming of the property are different: `FileTypeFilter` vs `FileTypeChoices`.

##### Suggestion:
 I'd suggest that both should have an identical name.

##### Problem:
The next thing is that all pickers only provide fast access to the NAME but not to the path. I don't know the reason for that but at least on Desktop you definitely want the file path. And here it gets annoying as the amount of code you'd have to write including error checks (as almost everything is declared as nullable) seems way too much:

Example of getting the selected folders:
```C#
var paths = await window.StorageProvider.OpenFolderPickerAsync(new FolderPickerOpenOptions() {
					AllowMultiple = true,
					Title = "Select folder"
				});

if (paths?.Count > 0) {

    List<string> results = new();
    foreach (var item in paths) {
        	if (item.TryGetUri(out Uri? uriPath))
        		results.Add(uriPath.LocalPath);
    }
}
```

##### Suggestion:
I'd suggest to provide direct access, for example:

```C#
public interface IStorageItem {

    /// <summary>
    /// Gets the local file path if there is one
    /// </summary>
    string? LocalPath { get; }

}
```

And for pickers where we can have multiple items I'd also suggest providing direct access:

```C#
public interface IStorageItem {

    /// <summary>
    /// Gets the local file paths of all items if there are any
    /// </summary>
    string[]? LocalPaths { get; }

}
```

Then we could reduce the amount of lines developers have to write down to:

```C#
var paths = await window.StorageProvider.OpenFolderPickerAsync(new FolderPickerOpenOptions() {
					AllowMultiple = true,
					Title = "Select folder"
				});

if (paths?.Count > 0) 
    List<string> results = new(paths);

```

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

API changes:
```diff
public interface IStorageProvider
{
...
+ Task<IStorageFile?> TryGetFileFromUri(Uri fileUri);
+ Task<IStorageFolder?> TryGetFolderFromUri(Uri folderUri);
+ Task<IStorageFolder?> TryGetWellKnownFolder(WellKnownFolder);
...
}

+ public static class StorageProviderExtensions
+ {
+   Task<IStorageFile?> TryGetFileFromPath(this IStorageProvider provider, string filePath)
+   Task<IStorageFolder?> TryGetFolderFromPath(this IStorageProvider provider, string folderPath)
+ }

public interface IStorageItem
{
...
- bool TryGetUri(out Uri path);
+ Uri Path { get; }
...
}

- public class BclStorageFile
- {
- }
- public class BclStorageFolder
- {
- }
- public class BclStorageProvider
- {
- }
```

One possible change, for the consistency with IStorageFolder.GetItemsAsync(), I can unify TryGetFileFromUri and TryGetFolderFromUri into TryGetItemFromUri method. Which later can be pattern matched depending on the runtime type.

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Yes, but breaking changes in between preview releases. Should be merged before RC.

## Fixed issues

Fixes partially #9876
Fixes #9029
Fixes #10034
Fixes #9014 (from discussion)

### CODE REVIEW

File: src/iOS/Avalonia.iOS/Storage/IOSStorageProvider.cs [hunk 1]
```diff
@@ -99,6 +100,40 @@ public async Task<IReadOnlyList<IStorageFile>> OpenFilePickerAsync(FilePickerOpe
             ? new IOSStorageFolder(url) : null);
     }
 
+    public Task<IStorageFile?> TryGetFileFromPath(Uri filePath)
+    {
+        // TODO: research if it's possible, maybe with additional permissions.
+        return Task.FromResult<IStorageFile?>(null);
```
Comment (maxkatz6) [inline src/iOS/Avalonia.iOS/Storage/IOSStorageProvider.cs:106]: My knowledge of iOS sandboxing ends here. And I wouldn't say it's safe to just create a new file/folder from the Uri, without checking for the permissions to the file/folder first. 
So instead, I made it to return null just like browser backend.

**Comment by workgroupengineering** [comment]
I think we should add an optional CancellationToken parameter to allow long operations to be canceled.

**Comment by maxkatz6** [comment]
@workgroupengineering it could make sense on open picker methods. But unfortunately, native OSs don't really provide this option. Except Android OS, where you can cancel activity.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028582-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028933-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029030-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029043-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029081-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**